### ISSUE TITLE
Early stopping and cross-validation

### ISSUE THREAD

**Issue by dleebrown** [dleebrown/ANNA#3]
Modify the training function to do more testing diagnostics during the training process. Specify a separate binary file to use as watchlist and define some sort of early stopping parameter that will stop training once error stops improving for some number of iterations.

**Comment by dleebrown** [comment]
Working on this. Complication due to using an input queue now. Will try changing batch_size to be a placeholder and then for cross validation just feeding in the entire xval data set and pulling a batch_size off the queue, where batch_size is just the size of the xval set. Will also need to modify add_to_queue to have a flag specifying whether or not to randomly select items or just select the first fetch_size stars, so that I can reuse the function to preprocesss all the xval set.

**Comment by dleebrown** [comment]
Completed. Implemented a second queue for the cross-validation set run in its own thread, and a master queue control op that switches between the input and cross-validation queues as needed. Also modified add_to_queue to be able to randomly select stars or just select the first fetch_size. 

Implemented optional early stopping. Will automatically use the cross-validation set for early stopping if one is provided, otherwise uses the current batch cost. Also changed the interval between calculating cross-validation cost/logging data to be an entry in the parameter file. Updated how number of iterations completed is passed between different calls of train_loop.
